gnosis-buyback/contracts/Buyback.sol

```diff
@@ -0,0 +1,539 @@
+pragma solidity ^0.5.2;
+
+import "@gnosis.pm/util-contracts/contracts/Token.sol";
+import "@gnosis.pm/dx-contracts/contracts/DutchExchange.sol";
+import "@gnosis.pm/dx-contracts/contracts/Oracle/PriceOracleInterface.sol";
+import "./SafeMath.sol";
+
+contract BuyBack {
+    using SafeMath for uint;
+
+    address owner;
+    DutchExchange public dx;
+
+    // Buyback config
+    struct Buyback {
+        address userAddress;
+        address sellToken;
+        address buyToken;
+        address burnAddress;
+        bool shouldBurnToken;
+        uint[] auctionIndexes; // This is a mapping of auction id to amount
+        uint tipAmount;
+        uint expires; // time when buyback expires if not invoked
+        bool hasExecutedBuyback;
+        bool claimedSellOrder;
+
+    }
+
+    Buyback[] public buybacks;
+    mapping(address => uint[]) public userBuybacks;
+
+    // maps user address to token address to total buyback amount
+    // helps to ensure that if you have multiple buybacks you can fulfil them
+    mapping(address => mapping(address => uint)) public userBuybackAmountTotal;
+
+    // maps user address to total buyback tips
+    mapping(address => uint) public userTipAmountTotal;
+
+    // mapping of the user address to token addresses & balance
+    mapping (address => mapping(address => uint)) public balances;
+
+    // mapping of user address to ether deposit value
+    mapping(address => uint) public etherBalance;
+
+    // mapping of buyback id to auction index to ductchx index for the auction index participated in
+    mapping(uint => mapping(uint => uint)) public dxAuctionIndex;
+    // mapping of buyback id to auction id to whether an auction has been clamied
+    mapping(uint => mapping(uint => bool)) internal alreadyClaimed;
+
+    // mapping of buyback id to auction index and auction amount
+    mapping(uint => mapping(uint => uint)) public auctionIndexWithAmount;
+
+
+    modifier onlyOwner () {
+        require(msg.sender == owner);
+        _;
+    }
+
+    /**
+     * @notice Create the buyback contract
+     * @param _dx Address of the dutch exchange
+     */
+    constructor(address _dx) public {
+        require(address(_dx) != address(0), "Invalid address");
+        dx = DutchExchange(_dx);
+        owner = msg.sender;
+    }
+
+    /**
+     * @notice addBuyBack
+    */
+    function addBuyBack(
+        address _buyToken,
+        address _sellToken, 
+        address _burnAddress, 
+        bool _burn, 
+        uint[] memory _auctionIndexes, 
+        uint[] memory _auctionAmounts,
+        uint _tipAmount,
+        uint _expires
+        ) public {
+        address _userAddress = msg.sender;
+
+        require(hasEnoughDeposit(_auctionAmounts, _sellToken, _userAddress), "user does not have enough deposit to create buyback");
+        require(hasEnoughTippingDeposit(_userAddress, _tipAmount), "user does not have enough tip balance");
```

Comment:
But, if the user adds for example:

```js
{
  ...
  _auctionIndexes: [35, 50],
  _auctionAmounts: [10, 10],
  _tipAmount: 5
}
```

- If the user deposits for tipping `5`, It will allow him to create the Buyback. 
- If a user executes the buyback for auction 35, it will get the tip
- If a user executes the buyback for auction 50, it will REVERT, cause there's not tip. This will break the promise and make the user to loose gas

Reply:
Fixed to allowing the user to have only one index per buyback since multiple buybacks can be created now